src/NumericalAlgorithms/Interpolation/IrregularInterpolant.cpp

```diff
@@ -0,0 +1,88 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "IrregularInterpolant.hpp"
+
+#include <array>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp"
+// IWYU pragma: no_forward_declare Tensor
+
+namespace {
+
+template <size_t Dim>
+Matrix interpolation_matrix(const Index<Dim>& mesh,
+                            const tnsr::I<DataVector, Dim>& points) noexcept;
+template <>
+Matrix interpolation_matrix(const Index<1>& mesh,
+                            const tnsr::I<DataVector, 1>& points) noexcept {
+  // For 1d interpolation matrix, simply return the legendre-gauss-lobatto
+  // matrix.
+  return Basis::lgl::interpolation_matrix(mesh[0], get<0>(points));
+}
+
+template <>
+Matrix interpolation_matrix(const Index<2>& mesh,
+                            const tnsr::I<DataVector, 2>& points) noexcept {
+  const std::array<Matrix, 2> matrices{
+      {Basis::lgl::interpolation_matrix(mesh[0], get<0>(points)),
+       Basis::lgl::interpolation_matrix(mesh[1], get<1>(points))}};
+
+  const auto number_of_points = get<0>(points).size();
+  Matrix result(number_of_points, mesh.product());
+  size_t s = 0;
```

Comment:
Is there a reason not to set this inside the first for loop? No need for scope leakage as far as I can see.